### PR TITLE
Fix neovim support and misc refactoring

### PR DESCRIPTION
**Pull request by johan-bjareholt**
Still kind of a WIP, not finished with the refactoring

### CODE REVIEW

**Comment by ahnlabb** [comment]
Looks really good! Most of this will have to be merged, but currently I'm hesitant for two reasons:
- From my profiling it looks like it's a 5 time performance decrease in vim-land
- The reason for the number_of_moves variable is that I've been experimenting with logging how active I was in vim by measuring my cpm. This data is not pre-merge friendly

**Comment by johan-bjareholt** [comment]
> From my profiling it looks like it's a 5 time performance decrease in vim-land

Huh did not even know that you could profile that, will look that up

> The reason for the number_of_moves variable is that I've been experimenting with logging how active I was in vim by measuring my cpm. This data is not pre-merge friendly

Ah, good point.
Will fix when I have the time!

**Comment by johan-bjareholt** [comment]
Performance didn't seem that bad, these are my results

nvim:
```
SCRIPT  /home/johan/Programming/activitywatch/aw-watcher-vim/plugin/activitywatch.vim
Sourced 1 time
Total time:   0.005025
 Self time:   0.000468
```

vim
```
SCRIPT  /home/johan/Programming/activitywatch/aw-watcher-vim/plugin/activitywatch.vim
Sourced 1 time
Total time:   0.001375
 Self time:   0.000670
```

**Comment by johan-bjareholt** [comment]
Added a startup performance script, since this is not interactive and doesn't do any heartbeats it's much faster though. Not really a good measure of performance but nice to have.

```
SCRIPT  /home/johan/Programming/activitywatch/aw-watcher-vim/plugin/activitywatch.vim
Sourced 1 time
Total time:   0.000524
 Self time:   0.000288
```

**Comment by ahnlabb** [comment]
That's indeed not a bad loading time, however `Poke` is called every time you move the cursor/write. That's the bottleneck that worries me. I'll post my profiling data tomorrow. It's probably fine but we should be very aware of it, any hint of sluggishness at high writing speeds is a big no-no among vim users.

**Comment by johan-bjareholt** [comment]
Ah, got it.

**Comment by johan-bjareholt** [comment]
> The reason for the number_of_moves variable is that I've been experimenting with logging how active I was in vim by measuring my cpm. This data is not pre-merge friendly

Also regarding this, calculating average CPM would be much easier to code in python with the solution I currently have here, if has bad performance though it's a no-go.

**Comment by ahnlabb** [comment]
It is probably fine, the time for each invocation is really short. I will do more rigorous profiling tomorrow:
- start profiler
- `profile func *Poke`
- emulate typist level wpm (50 - 80 wpm, i.e. 250 - 400 cpm)
- from the profiling data, calculate average time of Poke
- divide by time between keystrokes, the result should be much smaller than 1.

I'm guessing it is ~ 0.001 which seems more than acceptable

**Comment by ahnlabb** [comment]
My profiling results

dev/testing:
```
FUNCTION  <SNR>2_Poke()
Called 601 times
Total time:   0.142052
 Self time:   0.045745
```

master:
```
FUNCTION  <SNR>2_Poke()
Called 601 times
Total time:   0.041742
 Self time:   0.041162
```

At 300 cpm that is ~0.0012 of the time between each click spent in the Poke function in the dev/testing version which is perfectly fine.

The startup_performance.sh is commited to the repo but startup.profile is in the gitignore. If you think we should have the `startup_performance.sh` in the repo, move it to a `profiling` directory, add `profile.vim` and change `/dev/null` in that file to NONE (I believe that's what you want)

```
"vim -u NORC" can be used to skip these initializations without
reading a file.  "vim -u NONE" also skips plugins and syntax
highlighting.  -u
```

Have a look at my comments in the vim-file and merge when you feel like it

File: plugin/activitywatch.vim [hunk 1]
```diff
@@ -7,112 +7,72 @@ let g:loaded_activitywatch = 1
...
-        call s:Send("stop")
-        call s:CloseWatcher()
-endfunc
-
-function! s:CloseWatcher()
-        let s:vimwatcher_open = 0
+        if has('nvim')
+                call s:Send({'action': 'stop'})
+                call jobwait([s:vimwatcher_job], 5)
+                let s:vimwatcher_job = 0
```
Comment (ahnlabb) [inline plugin/activitywatch.vim]: Let `on_exit` handle this in nvim
EDIT: (Specifically, remove jobwait and let s:vimwatcher_job)

File: plugin/activitywatch.vim [hunk 2]
```diff
@@ -7,112 +7,72 @@ let g:loaded_activitywatch = 1
...
 
 function! s:Send(msg)
         if s:CheckStatus()
                 if has('nvim')
                         let l:json_msg = json_encode(a:msg)
-                        call jobsend(s:vimwatcher_job, l:json_msg)
+                        call jobsend(s:vimwatcher_job, l:json_msg . "\n")
                 else
-                        call ch_sendexpr(s:vimwatcher_job, a:msg)
+                        let l:json_msg = json_encode(a:msg)
```
Comment (ahnlabb) [inline plugin/activitywatch.vim]: Move this out of the if-else statement (cf. line 65)

**Comment by johan-bjareholt** [comment]
Now requires this aw-client PR for the reasons described in the PR: https://github.com/ActivityWatch/aw-client/pull/35

**Pull request merged**